kubelet: set terminationMessagePath perms to 0660

skrobul opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Currently, kubelet creates a world-readable and world-writeable empty files in `/var/lib/kubelet/pods/{podUID}/containers/{containerName}/{containerId}`. These are meant to be written by the process in containers when container is terminated.

Originally, this file was created with `0644`, then despite security concerns, it was changed to `0666` in https://github.com/kubernetes/kubernetes/issues/31839. This was completed to allow containers running as non-root to write termination messages. Later on, in 2019 this has been highlighted as a security vulnerability in Kubernetes Security Audit Report in https://github.com/kubernetes/kubernetes/issues/81116.

This PR changes termination log file mode to `0660` which is the best of both worlds - it removes world-writable file, yet still allows the container user and it's group to write the termination message.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Related (fixes only part) #81116

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Change world-accessible permissions to owner and group only read/write for files created by kubelet '/var/lib/kubelet/pods/{podUID}/containers/busysleep/{containerId}'. 
Termination message file (by default `/dev/termination-log`) can now be written only by the container user and it's group.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->

- [2019 Kubernetes Security Audit](https://github.com/kubernetes/sig-security/blob/6f1cec8878c705b67982e9b3bf3b52d6f19e17e0/sig-security-external-audit/security-audit-2019/findings/Kubernetes%20Final%20Report.pdf)

skrobul committed:
kubelet: set terminationMessagePath perms to 0660

Currently, kubelet creates a world-readable and world-writeable empty files in
`/var/lib/kubelet/pods/{podUID}/containers/busysleep/{containerId}`. These are
meant to be written by the process in containers when container is terminated.

Originally, this file was created with `0644`, then despite security concerns,
it was changed to `0666` in
https://github.com/kubernetes/kubernetes/issues/31839. This was completed to
allow containers running as non-root to write termination messages. Later on, in
2019 this has been highlighted as a security vulnerability in Kubernetes
Security Audir Report in https://github.com/kubernetes/kubernetes/issues/81116.

This commit changes termination log file mode to `0660` which is best of both
worlds - it removes world-writeable file, yet still allows the container's user
to write the termination message.

ehashman reviewed (commented):
/ok-to-test
/triage accepted
/priority important-longterm

skrobul commented:
/retest

skrobul commented:
/remove-lifecycle stale

skrobul commented:
/remove-lifecycle staleFrom: Kubernetes Triage Robot ***@***.***>Sent: Wednesday, 5 October 2022 13:02To: kubernetes/kubernetesCc: Marek Skrobacki; MentionSubject: Re: [kubernetes/kubernetes] kubelet: set terminationMessagePath perms to 0660 (PR #108076)
The Kubernetes project currently lacks enough contributors to adequately respond to all issues and PRs.
This bot triages issues and PRs according to the following rules:

After 90d of inactivity, lifecycle/stale is applied
After 30d of inactivity since lifecycle/stale was applied, lifecycle/rotten is applied
After 30d of inactivity since lifecycle/rotten was applied, the issue is closed

You can:

Mark this issue or PR as fresh with /remove-lifecycle stale
Mark this issue or PR as rotten with /lifecycle rotten
Close this issue or PR with /close
Offer to help out with Issue Triage

Please send feedback to sig-contributor-experience at kubernetes/community.
/lifecycle stale

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>
[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/kubernetes/kubernetes/pull/108076#issuecomment-1268280916",
"url": "https://github.com/kubernetes/kubernetes/pull/108076#issuecomment-1268280916",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

haircommander commented:
I feel like there should be a test to verify both cases. Then, we can stop flip flopping and ensure this is the behavior we want

haircommander commented:
/remove-lifecycle rotton

skrobul commented:
/remove-lifecycle rotten

> I feel like there should be a test to verify both cases.

@haircommander I didn't include the test because I feel like this would be testing if operating system sets the exact permission it is asked for. Can you please elaborate on what would you expect such test to do?

haircommander commented:
I suppose I had thought because it was using osInterface it could be spoofed and unit tested. Though, the function around it looks complicated enough that setting it up may be a pain. I am okay without a test

/lgtm

rata commented:
> This PR changes termination log file mode to `0660` which is the best of both worlds - it removes world-writable file, yet still allows the container's user to write the termination message.

@skrobul Just curious, the termination-log owner is not root:root? Can you point where is the user/group of that file changed to match the container user?

skrobul commented:
> > This PR changes termination log file mode to `0660` which is the best of both worlds - it removes world-writable file, yet still allows the container's user to write the termination message.
> 
> @skrobul Just curious, the termination-log owner is not root:root? Can you please point where is the user/group of that file changed to match the container user? Do you have that handy?

The owner of the /dev/termination-log by default is  `root:root` and as far as I know it is not changed to match the container. I guess for running non-root containers, one can [specify custom `terminationMessagePath`](https://kubernetes.io/docs/tasks/debug/debug-application/determine-reason-pod-failure/#customizing-the-termination-message) pointing to the file owned by any user. 

So in summary, I was wrong - should have said "yet still allows root to write the termination message".

rata commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -411,7 +411,7 @@ func (m *kubeGenericRuntimeManager) makeMounts(opts *kubecontainer.RunContainerO
 			// ~umask". But we want to make sure the specified mode is used
 			// in the file no matter what the umask is.
-			if err := m.osInterface.Chmod(containerLogPath, 0666); err != nil {
+			if err := m.osInterface.Chmod(containerLogPath, 0660); err != nil {
```
The PR mentions the terminationMessagePath that is deined a few lines below: https://github.com/kubernetes/kubernetes/pull/108076/files#diff-d86ffc24d8e4c9deafd3e39e8f4ee3546ee46a31dcea8ea6526443b3f4bffa28R420

How is it that changing the `containerLogPath` affects the `terminationMessagePath`?

skrobul replied:
ok, so my goal is to get rid of the world writeable file regardless of how is it called. The error message on line 406 suggests that `termination-log file` is the same as `containerLogPath` - is it not?

The `terminationMessagePath` as I understand it is an absolute path that is used when mounting the volume and
it has not much to do with what is known as the concept of termination message despite the name. In fact, even the error message on 415 refers to `containerLogPath` rather than `terminationMessagePath`. 

At this point I am bit confused about this code - if that's not the line responsible for creating the empty world writable files like this:

```
# ls -la /var/lib/kubelet/pods/fa852268-2261-4ba4-9aaf-a4eaa98d1dfe/containers/kube-proxy/a62b249a
-rw-rw-rw- 1 root root 0 Mar 20 12:30 /var/lib/kubelet/pods/fa852268-2261-4ba4-9aaf-a4eaa98d1dfe/containers/kube-proxy/a62b249a
```

then what is?

rata replied:
It might very well be the same, I'm unsure. But some explanation on the commit msg would be helpful so it is easier to review, you probably figured that out when writing the PR :)

rata commented:
> The owner of the /dev/termination-log by default is `root:root` and as far as I know it is not changed to match the container. I guess for running non-root containers, one can [specify custom `terminationMessagePath`](https://kubernetes.io/docs/tasks/debug/debug-application/determine-reason-pod-failure/#customizing-the-termination-message) pointing to the file owned by any user.

@skrobul this doesn't seem the case. If you use a custom log-path, in my limited testing, this just changes the destionation of the bind mount. But a new bind-mount is created in that path, so the file even if existing in the container image, is cleared by the mount on top (I've tested this just using the pod from the link you posted, making the pod command just sleep infinity and changing the terminationMessagePath to /bin/grep that exists on the debian base image).

It seems this restricts the feature to be usable only in containers running as root. Or am I missing something?

skrobul commented:
> @skrobul this doesn't seem the case. If you use a custom log-path, in my limited testing, this just changes the destionation of the bind mount. But a new bind-mount is created in that path, so the file even if existing in the container image, is cleared by the mount on top (I've tested this just using the pod from the link you posted, making the pod command just sleep infinity and changing the terminationMessagePath to /bin/grep that exists on the debian base image).

> It seems this restricts the feature to be usable only in containers running as root. Or am I missing something?

To be honest - I do not remember. If it's indeed changing just a mount path, then you are absolutely correct - that path will still be owned by root and non-root users will not be able to write to it. 

I'll setup my dev environment again and try to work on this as soon as I can. It sounds like it would make sense to ensure that we change ownership to the same UID as the container's UID.

rata commented:
> To be honest - I do not remember. If it's indeed changing just a mount path, then you are absolutely correct - that path will still be owned by root and non-root users will not be able to write to it.

The code looks like that and I've tested it already, it does that :)

> I'll setup my dev environment again and try to work on this as soon as I can. It sounds like it would make sense to ensure that we change ownership to the same UID as the container's UID.

Yes, but I'm not sure how easy it is to get that. We have the container info here, but this might be set in the pod or the container image too. Having a look on how clumsy that would be seems great.

Once you know the different options, I think you should join sig-node, add your topic to the agenda and ask there.

skrobul commented:
/test ?

skrobul commented:
@rata 

>  I've tested it already, it does that :)

Interestingly there seems to be an [e2e test](https://github.com/kubernetes/kubernetes/blob/0c62b122c02bff9131b6db960042150a3638d3f3/test/e2e/common/node/runtime.go#L190-L209) that already covers this scenario. The green checks in this PR seem to indicate that these tests were successful. Any ideas why that might be?

rata commented:
Interesting... Maybe I missed something? My gut feeling is that this change is not enough to remove permissions for others, but no idea really. I think you will need to run this locally, see what permissions are set for the host path and the container path and see if writing it as a non-root user works and why :-)

skrobul commented:
Turns out that mentioned e2e test specifies the non-root user, but it does not override the default group of root which results in normal user to still be able to write the message through group privileges. This is obviously not a desired behaviour here, so I am working on better solution but at least now I understand why this test is not failing.

skrobul committed:
node runtime test: use non-root group

skrobul committed:
kuberuntime: change ownership of the termination-log file

The default mode of the termination-log has been changed to 660 in
earlier commit. The intention of this change was to remove the
world-writable access to this file for security reasons. Unfortunately,
before this commit the termination-log was always owned by the root
user, regardless of what user container runs as.

Following this change, the ownership of termination-log file (be it
default /dev/termination-log or any custom specified by
TerminationMessagePath) will be changed to the uid and gid configured in
SecurityContext on the pod or container level.

rata commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -441,10 +441,33 @@ func (m *kubeGenericRuntimeManager) makeMounts(opts *kubecontainer.RunContainerO
 			}
 
+			var containerUid, containerGid int
+			if container.SecurityContext.RunAsUser != nil {
+				containerUid = int(*container.SecurityContext.RunAsUser)
+			} else if pod.Spec.SecurityContext.RunAsUser != nil {
+				containerUid = int(*pod.Spec.SecurityContext.RunAsUser)
+			} else {
+				containerUid = 0
```
This probably doesn't work if the user is specified in the Dockerfile.

I _think_ there are other parts of Kubernetes (something to check if it doesn't run as root? Something else too, but I can't remember now) that contemplate all, not just this.

Maybe it is a reasonable trade-off, though, to only check that. I don't know

skrobul replied:
> This probably doesn't work if the user is specified in the Dockerfile.

That is correct. I think it would be reasonable to expect end-user to configure SecurityContext explicitly in such case, but if there is a way to determine this automatically it would be better. I have not found anything that would allow to peak into the container image. The closest I was able to find was [`DetermineEffectiveRunAsUser(...)`](https://github.com/kubernetes/kubernetes/blob/d89d5ab2680bc74fe4487ad71e514f4e0812d9ce/pkg/securitycontext/util.go#L123), but it still only derives the information from the declared pod/container security context, not the image.

skrobul commented:
/retest

skrobul commented:
/retest

bart0sh commented:
/retest

bart0sh commented:
/cc
@skrobul please fix CI test failures, thanks.

skrobul committed:
kuberuntime: add nil check for SecurityContext

skrobul commented:
/retest

skrobul commented:
/retest

rata reviewed (commented):
The code looks reasonable _assuming this is what we want_. And tests are of course missing :)

However, I don't see much activity here to know what sig-node thinkgs about this. I'd recommend you, again, to join sig-node to have their interest and also ask if this should have a (very simple) KEP and feature gate or not. I think it probably should have.

If @SergeyKanzhelev or someone from sig-node can have a look and answer here, without the need to join sig-node, maybe that is simpler for you @skrobul ? :)

skrobul commented:
> I'd recommend you, again, to join sig-node to have their interest 

According to the [docs](https://kubernetes.io/docs/contribute/participate/) in order to become member, I need to show "multiple contributions" before being added to the organisation. This is my very first contribution, so not exactly there yet.

rata commented:
@skrobul wait, that is mixing several different things. First, that link is for sig-docs, that is the Single Interest Group (SIG) about documentation. I'm talking about sig-node.

Secondly, you don't need to join any github org to join the weekly meetings. You can find the sig-node info here: https://github.com/kubernetes/community/tree/master/sig-node. Anyone can join and discuss the topics they want.

To add topics to the agenda, you just need to subscribe to the sig-node mailing list, and then you add your topic to the next weeks agenda (if there isn't one created, you can create one). Then, you join at appropiate date and time on zoom, discuss with the community the problem and solution you propose, they will tell you if a KEP is needed, a feature gate, etc. (my guess is that a very lightweight KEP will be needed. But hopefully it isn't, let's see!)

skrobul commented:
> Secondly, you don't need to join any github org to join the weekly meetings. You can find the sig-node info here: https://github.com/kubernetes/community/tree/master/sig-node. Anyone can join and discuss the topics they want.

ohh that makes more sense now, thank you! I will do some reading and join soon. Once again thank you for taking time to write that up, appreciate it.

marquiz commented:
@skrobul are you still working on this? What was outcome of the discussion in SIG Node meeting (I think I missed that meeting myself)?

SergeyKanzhelev commented:
> @skrobul are you still working on this? What was outcome of the discussion in SIG Node meeting (I think I missed that meeting myself)?

Link: https://docs.google.com/document/d/1Ne57gvidMEWXR70OxxnRkYquAoMpt56o75oZtg-OeBg/edit#bookmark=id.8ats7bythcn4 There is a recording available.

I cannot check recording now, but from what I recollect we discussed that there might be scenarios that can be broken by this. Something around external tool writing termination message about oom kill or similar.

I think suggestion was that it will be a breaking change and we will need to run the KEP process to work ensure we did a due diligence.

But my recollection may be faulty.

marquiz commented:
Thanks @SergeyKanzhelev. I quickly checked the recording and you indeed suggested writing a KEP. I think @rata was inclined in the same direction. My feeling about this is similar: KEP would allow exploring potential risks and alternative solutions in a well-defined manner plus prevent breaking existing users (with feature gate)

/cc @dchen1107

skrobul commented:
> @skrobul are you still working on this?

No, I have done everything I could at this stage

>  What was outcome of the discussion in SIG Node meeting (I think I missed that meeting myself)?

The outcome as far as I remember was that:
- KEP is created for this. I did that about two months ago in https://github.com/kubernetes/enhancements/pull/4143/files and I did not see any responses.
- @dchen1107 agreed to review this PR

skrobul commented:
/remove-lifecycle stale

skrobul commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -441,10 +441,33 @@ func (m *kubeGenericRuntimeManager) makeMounts(opts *kubecontainer.RunContainerO
+			} else if pod.Spec.SecurityContext != nil && pod.Spec.SecurityContext.RunAsUser != nil {
+				containerUid = int(*pod.Spec.SecurityContext.RunAsUser)
+			} else {
+				containerUid = 0
```
as far as I understand the `RunAsNonRoot` does not influence the UID selection, it merely enables a validation that it's non-root, is that not the case?

bart0sh replied:
That's true, but if this change would set UID to 0 (root) even if RunAsNonRoot is true. It's may be ok, but looked suspicious to me. What would happen if container runs as non-root, but containerLogPath owner is root?

skrobul replied:
The container would not run at all if the `RunAsNonRoot` was set to true, but the `pod.Spec.SecurityContext.RunAsUser` was set unset. In other words, as far as I understand there is no way for the container to start as non-root without changing the `RunAsUser` either on container or pod level. What am I missing?

bart0sh replied:
I'm not sure about it. I thought that `RunAsNonRoot` and `RunAsUser` are independent options. By default user UID and/or username is taken from the image.

bart0sh replied:
Here is the code I'm talking about: https://github.com/kubernetes/kubernetes/blob/release-1.29/pkg/kubelet/kuberuntime/kuberuntime_container.go#L326

skrobul commented:
/remove-lifecycle rotten

skrobul commented:
/remove-lifecycle rotten

dims commented:
/test all

dims commented:
/easycla

dims commented:
/assign @SergeyKanzhelev

skrobul committed:
kuberuntime: linter updates
